src/args/types.mth

```diff
@@ -0,0 +1,107 @@
+module(args.types)
+
+import(prelude)
+import(data.maybe)
+import(data.str)
+
+import(args.parse)
+
+table(ArgumentParser)
+field(ArgumentParser~options, ArgumentParser, Maybe(List(ArgpOption)))
+field(ArgumentParser~parser, ArgumentParser, Maybe([ +ArgumentParser a Maybe(Str) ArgpOptionType -- +ArgumentParser a ]))
+field(ArgumentParser~args-doc, ArgumentParser, Maybe(Str))
+field(ArgumentParser~doc, ArgumentParser, Str)
+def(ArgumentParser.new, Maybe(List(ArgpOption)) Maybe([ +ArgumentParser a Maybe(Str) ArgpOptionType -- +ArgumentParser a ]) Maybe(Str) Str -- ArgumentParser,
+  ArgumentParser.alloc!
+  dup dip(~doc !)
+  dup dip(~args-doc !)
+  dup dip(~parser !)
+  dup dip(~options !)
+)
+
+def(ArgumentParser.parser, ArgumentParser -- Maybe([ +ArgumentParser a Maybe(Str) ArgpOptionType -- +ArgumentParser a ]), ~parser @)
+
+table(ArgumentParserState(a))
+field(ArgumentParserState~argv, ArgumentParserState, List(Str))
+field(ArgumentParserState~program-name, ArgumentParserState, Str)
+field(ArgumentParserState~current-argv-index, ArgumentParserState, Int)
+field(ArgumentParserState~current-positional-arg, ArgumentParserState, Int)
+field(ArgumentParserState~currently-parsing-option, ArgumentParserState, Bool)
+field(ArgumentParserState~current-option-type, ArgumentParserState, Maybe(ArgpOptionType))
+field(ArgumentParserState~current-option-option, ArgumentParserState, Maybe(ArgpOption))
+field(ArgumentParserState~current-error, ArgumentParserState, Maybe(ArgumentParsingError))
+field(ArgumentParserState~arguments, ArgumentParserState, a)
```

Comment:
The free type variable in the field type should not be allowed here. I opened issue #197.
This is also an issue with the `ArgumentParser~parser` field.

The way around these issues, within the type system as it currently exists, is to use `data` everywhere instead of `table`. I understand that's a big ask for this PR. So for now maybe we can "fix" it by replacing all the `a` with a concrete type? It limits the utility of this library for non-compiler work, but it's better to fix this issue conservatively than to work with unsound types.